nemo/collections/nlp/models/dialogue/intent_bio_type_classification_model.py

```diff
@@ -0,0 +1,844 @@
+        
+        # calculate accuracy metrics for intents and slot reporting
+        # intents
+        intent_preds = torch.argmax(intent_logits, axis=-1)
+        self.intent_classification_report.update(intent_preds, intent_labels)
+        
+        # slots
+        loss_mask = loss_mask > 0.5
+        bio_slot_preds = torch.argmax(bio_slot_logits, axis=-1)
+        self.bio_slot_classification_report.update(bio_slot_preds[loss_mask], bio_slot_labels[loss_mask])
+
+        mention_loss_mask = mention_loss_mask > 0.5
+        slot_preds = torch.argmax(slot_logits, axis=-1)
+        self.slot_classification_report.update(slot_preds[mention_loss_mask], bio_mention_labels[mention_loss_mask])
+
+        slot_similarity_preds = torch.argmax(dot_product_score_log_softmax, axis=-1)
+        self.slot_similarity_classification_report.update(slot_similarity_preds[mention_loss_mask], bio_mention_labels[mention_loss_mask])
+
+        return {
+            'val_loss': val_loss,
+            'intent_tp': self.intent_classification_report.tp,
+            'intent_fn': self.intent_classification_report.fn,
+            'intent_fp': self.intent_classification_report.fp,
+            'bio_slot_tp': self.bio_slot_classification_report.tp,
+            'bio_slot_fn': self.bio_slot_classification_report.fn,
+            'bio_slot_fp': self.bio_slot_classification_report.fp,
+            'slot_tp': self.slot_classification_report.tp,
+            'slot_fn': self.slot_classification_report.fn,
+            'slot_fp': self.slot_classification_report.fp,
+            'slot_similarity_tp': self.slot_similarity_classification_report.tp,
+            'slot_similarity_fn': self.slot_similarity_classification_report.fn,
+            'slot_similarity_fp': self.slot_similarity_classification_report.fp,
+            'intent_preds': intent_preds,
+            'intent_labels': intent_labels,
+            'bio_slot_preds': bio_slot_preds,
+            'bio_slot_labels': bio_slot_labels,
+            'slot_preds': slot_preds,
+            'slot_labels': slot_labels,
+            'slot_similarity_preds': slot_similarity_preds,
+            'input': input_ids,
+            'subtokens_mask': subtokens_mask,
+        }
+
+    def get_entity_description_score(self, mention_hidden_states_pad):
+        """
+        Implement 3.7 entity description score from ReFinED paper
+        Args: 
+            mention_hidden_states_pad: mention embedding of size (batch_size * max_num_mention * hidden_state_dim)
+            self.description_embeddings: description embeddings of size (num_slot_class * shared_embedding_dim)
+            
+        Returns:
+            dot product score matrix of given mention embedding and description embedding (batch_size * max_num_mention * num_slot_class)
+        
+        """
+        if next(self.description_projection_mlp.parameters()).device != self.device:
```

Comment:
Please write a comment to explain what is done here (make sure things are on the same device).